### PR TITLE
No-chop merge and delete

### LINKED ISSUE #1327: No-chop merges
Related to #1050 
Supersedes https://issues.apache.org/jira/browse/ACCUMULO-3235 and https://issues.apache.org/jira/browse/ACCUMULO-4118

Merges currently require a special "chop" major compaction, which can be slow. The idea is simple: in the metadata entries for files, we track the range in use for that file with a notation.

This additional metadata would be used to merge tablets without re-writing data. Only this extra metadata would be updated (to a narrower range when split, and possibly collapsed if merging adjacent ranges). If a range is missing for a file, the tablet's range will be used instead. This simplifies things for backwards compatibility, and means we don't need to include a redundant range if the tablet isn't splitting/merging.

In order to achieve this, we will need a mechanism to serialize the range information into the file field of the tablet metadata, and we will need to ensure that all reads of the file are confined to the range serialized. Some pitfalls to look out for: files opened multiple times with different ranges, use of block cache, offline file reads, handling duplicate entries for a file with different ranges as the result of a merge, garbage collection of files, etc.

### PR DESCRIPTION
**Pull request by cshannon**
This PR is to merge in the no-chop merge and no-chop/no-split feature into main. There are still some follow on tasks to do which I will list and create issues for but the main functionality is complete and now that 3.0 has been release I think it's time to get what is here reviewed and merged into main so we can continue on.

The goal here is to prevent having to run chop compactions when merging (or deleting) tablets which can be quite slow. The other goal is to get rid of having to split tablets during deletions to simplify the manager code. Instead, the changes here allow RFiles to be fenced off by by valid ranges so that we don't need to compact/split anymore and when scanning only the valid range is visible. Files can then be cleaned up later with the normal compaction process.

**Notes:** 

1. This PR squshes all the commits from the [no-chop-merge](https://github.com/apache/accumulo/tree/no-chop-merge) branch into one commit to make it much easier to review. I left that original branch alone for now and pushed this to a different branch.
2. I plan to create a wiki page or website post at some point with full details of how no-chop merge works and the changes (as it would be quite large) but the highlights are below for reviewing the PR:

### Description of Changes:

**TabletFile**

Tablet files used to be unique by just the path. Now a TabletFile also includes a Range. This allows Accumulo to track the same file more than one time, each with a different associated range. So the same file can have several valid ranges and be treated as different files. The benefit here is most of the TabletFile handling code can stay the same. If a Range is not provided (null start/end) then the entire file is visible and this is how backwards compability is handled.

**RFiles**

There is now a new RFile iterator/reader that can fence files by taking a collection of ranges and only returning rows that are part of the given range when scanning. Rows outside of the range are skipped and not visible.

**Metadata**

The File metadata that is stored has been changed to now include a range. Originally each file just contained the path in the Data File column. Now this has been changed so that the value is Json and includes both a file and a range. This means that the same file can be stored more than one time in metadata as the combination of file and range makes it unique.

**Merge Strategy**

During a merge, instead of compacting tablets, each tablet that is being merged will instead have its files fenced by the tablet range. This is a quick operation as this just requires going through the files in metadata and adding the range to the entry and then adding that entry to the target tablet where files are being merged to . If an existing range already exists for the file when processing then that file has its range clipped with the tablet range to get the valid overlapping range. Each file will only have 1 range created.

**Delete Strategy**

Deletes are a little more complicated because besides chop compactions when there are shared boundaries (tablets are stretched to cover the deleted range so we need to compact), we also need to handle splits. Normally splits are first done at the deletion boundaries so entire tablets can be deleted but now we can also fence off tablets instead of splitting. Deletion now works by figuring out the first/last tablets in the deletion range and instead of splitting them or compacting the algorithm will fence the tablets (if needed) and create 0, 1 or 2 ranges. Then any tablets that exist in the middle are still just deleted like before.

When the deletion range overlaps part of the first tablet or part of the last tablet we need to create 1 valid range for the files in those tablets (the part of the files outside the deletion range and not being deleted). This is done instead of a split. If a deletion range is entirely inside of a single tablet then we need to create two ranges (two files are added for each existing file) as we need to fence all the files by the valid range before and after the delete range.

**Future Issues to be Done**

1. All of the existing chop compaction related code can be removed (or deprecated if public API). There's a lot to remove and would make the PR much larger so I figured it was best to do a follow on.
3. We still need some more testing to make sure all edge cases are handled
4. Upgrade code needs to be written once the changes are approved
5. There might be some simpler things we could do to improve things as a follow on. One example @keith-turner mentioned was maybe instead of deleting tablets during deletion we could just remove the files and call merge later. We may also be able to simplify the metadata updates into 1 method/loop vs 2 for deletes. I plan to explore some of this going forward.
6. All of the changes will need to be merged/refactored into Elasticity branch.

**Included PRs:**

https://github.com/apache/accumulo/pull/3418, https://github.com/apache/accumulo/pull/3480, https://github.com/apache/accumulo/pull/3614, https://github.com/apache/accumulo/pull/3621, https://github.com/apache/accumulo/pull/3640, https://github.com/apache/accumulo/pull/3710, https://github.com/apache/accumulo/pull/3728

This closes #1327

### CODE REVIEW

**Comment by cshannon** [comment]
I restarted the full IT build with this new PR

**Comment by cshannon** [comment]
I merged in main to get #3760 to fix the tests, I just re-kicked the full IT again.

**Comment by cshannon** [comment]
> I merged in main to get #3760 to fix the tests, I just re-kicked the full IT again.

Full IT build finished and passed.

**Review by keith-turner — COMMENTED** [review]
I have been running this on a single node instance and experimenting with it, its really neat and is working really well. I created [this gist](https://gist.github.com/keith-turner/08cb06d84d3040fb60b2bc058b9e9a5d) showing some of the things I tried.   One problem I ran into is there is no easy way to see the ranges, I think we need a follow on issue for this. Need something in place for debugging purposes that makes it easy to see the ranges.

I manually exported and imported a table where the files had non infinite and it worked, the ranges made it. Do you know if there are test for this?  If not I think we should open a follow on issue to create test for this.

I manually cloned a table that had files with non infinite ranges and that worked great.  If that does not have test, we probably want to open a follow on issue for that.

Noticed MergeState still has WAITING_FOR_CHOPPED state, is that something you were thinking of removing as a follow on?  Have you created any issues for known follow on work yet?

Still looking at the code changes again.

**Comment by cshannon** [comment]
> I have been running this on a single node instance and experimenting with it, its really neat and is working really well. I created [this gist](https://gist.github.com/keith-turner/08cb06d84d3040fb60b2bc058b9e9a5d) showing some of the things I tried. One problem I ran into is there is no easy way to see the ranges, I think we need a follow on issue for this. Need something in place for debugging purposes that makes it easy to see the ranges in the metadata table.
> 

I can create a follow on issue. For debug logging (if we don't already have the `StoredTabletFile` object loaded) we could deserialize the metadata back into either a `StoredTabletFile` or just to the json object that is used to serialize which is `TabletFileCqMetadataGson` and use that to print it as human readable form. Reading it back into `TabletFileCqMetadataGson` would be quicker and more efficient if we just wanted the Range than loading an entire `StoredTabletFile`.

For seeing the ranges when writing ITs the simplest thing was to deserialize the metadata entry back into the entire `StoredTabletFile` object and then print the range object and other things. I did this for all of the ITs I modified for merge/delete so I could easily see what was stored. The helper method I used to print the metadata with human readable ranges is part of this PR in [FileMetadata](https://github.com/apache/accumulo/blob/ab3e55147fb11e4df55c7a47680fa3b5f85cee4a/test/src/main/java/org/apache/accumulo/test/util/FileMetadataUtil.java). I just added a log statement where I print the Range and other info: https://github.com/apache/accumulo/blob/ab3e55147fb11e4df55c7a47680fa3b5f85cee4a/test/src/main/java/org/apache/accumulo/test/util/FileMetadataUtil.java#L61-L63

This outputs like the following: `Extent: 5<<; File Name: F000000c.rf; Range: (-inf,m%00; : [] 9223372036854775807 false); Entries: 2, Size: 212`

> I manually exported and imported a table where the files had non infinite ranges and it worked, the ranges made it. Do you know if there are test for this? If not I think we should open a follow on issue to create test for this.
> 
I haven't specifically done any testing with import/export so I can create a follow on for testing. My assumption is that it should work (along with bulk import) due to everything is just treated as a StoredTabletFile in the rest of the code base and the fact that it has a range shouldn't really matter but we should definitely test it.
> I manually cloned a table that had files with non infinite ranges and that worked great. If that does not have test, we probably want to open a follow on issue for that.
> 
Same thing here, I don't think there's a test (at least I didn't write one). I am thinking I can create an overall top level issue for adding testing for no-chop merge and we can just keep a running list of things to add and add PRs to cover each case as there's likely several things we want to test with ranges.
> Noticed MergeState still has WAITING_FOR_CHOPPED state, is that something you were thinking of removing as a follow on? Have you created any issues for known follow on work yet?
> 
Right, I figured I would just clean up all "chopped" related things in a follow on. I started to work on removing all of that stuff, like `WAITING_FOR_CHOPPED` from MergeState, removing things from Thrift, etc and the diff was starting to hit another 1000 lines of code changes to remove all the chop compaction code. So I figured for now I'd just leave it all as it doesn't hurt anything (just not used) and I was going to create an issue and follow on PR to remove all of it at one time.

File: core/src/main/java/org/apache/accumulo/core/metadata/StoredTabletFile.java [hunk 1]
```diff
@@ -118,11 +148,115 @@ public String toString() {
...
+
+  private static byte[] encodeRow(final Key key) {
+    final Text row = key != null ? key.getRow() : null;
+    if (row != null) {
+      try (DataOutputBuffer buffer = new DataOutputBuffer()) {
+        row.write(buffer);
+        return buffer.getData();
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
```
[the remark this answers is not in the document]
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/StoredTabletFile.java]: @keith-turner - Thanks for the suggestion, I made the changes with the serialization to shrink how much is written and also the html escaping. I had also notice the long out put as well a while back and meant to go back and figure out why but I forgot because the last few weeks I've been testing with deserializing back into the Range object so it was human readable (as I noticed elsehwere) .

[the remark this answers is not in the document]
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/StoredTabletFile.java]: As a side note I forgot to include co-author tags for @keith-turner and @ctubbsii when I made this new PR and squashed the previous branch. I will try and remember to make sure they are included whenever we squash this PR and merge it into main.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/compactions/ExternalCompactionJob.java [hunk 2]
```diff
@@ -76,7 +76,7 @@ public TExternalCompactionJob toThrift() {
 
     List<InputFile> files = jobFiles.entrySet().stream().map(e -> {
       var dfv = e.getValue();
-      return new InputFile(e.getKey().getNormalizedPathStr(), dfv.getSize(), dfv.getNumEntries(),
+      return new InputFile(e.getKey().getMetadata(), dfv.getSize(), dfv.getNumEntries(),
```
Comment (keith-turner) [inline server/tserver/src/main/java/org/apache/accumulo/tserver/compactions/ExternalCompactionJob.java:79]: As a follow on it would be good to have an IT that forces an external compaction on a tablet that has files with ranges.  Construct the situation such that if the external compaction does not honor the ranges it would end up creating a file with more data than it should have.

File: test/src/main/java/org/apache/accumulo/test/functional/MergeIT.java [hunk 3]
```diff
@@ -102,6 +124,306 @@ public void mergeSize() throws Exception {
     }
   }
 
+  @Test
+  public void noChopMergeTest() throws Exception {
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/functional/MergeIT.java:128]: The test that were added are nice and the comments make them easier to understand.

File: server/base/src/main/java/org/apache/accumulo/server/fs/VolumeUtil.java [hunk 4]
```diff
@@ -53,29 +54,29 @@ public static String removeTrailingSlash(String path) {
   }
 
   public static Path removeTrailingSlash(Path path) {
-    String pathStr = path.toString();
+    String pathStr = Objects.requireNonNull(path).toString();
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/fs/VolumeUtil.java:57]: There are volume replacement ITs may be worthwhile having those exercise replacing a volume where there are tablets that are using that volume with non infinite ranges.  This could be another follow in testing issue.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java [hunk 5]
```diff
@@ -1556,5 +1561,230 @@ private void setInterruptFlagInternal(AtomicBoolean flag) {
...
+    private final FileSKVIterator reader;
+    protected final Range fence;
+    private final Key fencedStartKey;
+    private final Key fencedEndKey;
+
+    public FencedFileSKVIterator(FileSKVIterator reader, Range fence) {
+      this.reader = Objects.requireNonNull(reader);
+      this.fence = Objects.requireNonNull(fence);
+      this.fencedStartKey = fence.getStartKey();
+      this.fencedEndKey = getEndKey(fence.getEndKey());
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:1588]: I would suggest moving this computation of the fencedEndKEy to the getLastKey() method.  Reason is that not every scan will need this.  Only things like splitting a tablet will use it.  So I suspect its being frequently computed but not frequently used.  Maybe the fencedEndKey instance var could be dropped.

```suggestion
```

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java [hunk 6]
```diff
@@ -1556,5 +1561,230 @@ private void setInterruptFlagInternal(AtomicBoolean flag) {
...
+
+    @Override
+    public FileSKVIterator getIndex() throws IOException {
+      return new FencedIndex(reader.getIndex(), fence);
+    }
+
+    @Override
+    public FileSKVIterator getSample(SamplerConfigurationImpl sampleConfig) {
+      final Reader sample = reader.getSample(sampleConfig);
+      return sample != null ? new FencedReader(sample, fence) : null;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:1769]: As a follow on could have test of sampling that covers case of files with non infinite ranges.

File: core/src/main/java/org/apache/accumulo/core/gc/ReferenceFile.java [hunk 7]
```diff
@@ -49,16 +67,16 @@ public TableId getTableId() {
   }
 
   @Override
-  public String getMetadataEntry() {
-    return metadataEntry;
+  public String getMetadataPath() {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/gc/ReferenceFile.java:70]: Looking at the javadoc for [Reference.getMetadataPath()](https://github.com/apache/accumulo/blob/f501dca59ae47179c6f62a134e44b27a3d34f652/core/src/main/java/org/apache/accumulo/core/gc/Reference.java#L39-L45) it writes about this data being exactly what is stored in the metadata table.  That is not technically true, what it stored in the metadata table is the json.  However I don't think it needs to be exactly what is stored in the metadata table.  Need to figure out if there is need for the exact metadata info or if the the javadoc needs to be updated.   I suspect the javadoc needs to be updated, but not completely sure.  Could open a follow on issue to investigate this further.

**Review by keith-turner — COMMENTED** [review]
Still looking at this, I have done 70 of 78 files.  Posting the comments I have made so far.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/ExternalCompactionMetadata.java [hunk 8]
```diff
@@ -146,7 +144,7 @@ public static ExternalCompactionMetadata fromJson(String json) {
     return new ExternalCompactionMetadata(
         jData.inputs.stream().map(StoredTabletFile::new).collect(toSet()),
         jData.nextFiles.stream().map(StoredTabletFile::new).collect(toSet()),
-        new ReferencedTabletFile(new Path(jData.tmp)), jData.compactor,
+        StoredTabletFile.of(jData.tmp).getTabletFile(), jData.compactor,
```
[the remark this answers is not in the document]
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/ExternalCompactionMetadata.java:147]: I create #3768 to handle upgrades and I created a task list to track all of the different things that need upgrading. I added this to the task list so we don't forget.

File: test/src/main/java/org/apache/accumulo/test/util/FileMetadataUtil.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+          files.put(file, dfv);
+          log.debug("Extent: " + tabletMetadata.getExtent() + "; File Name: " + file.getFileName()
+              + "; Range: " + file.getRange() + "; Entries: " + dfv.getNumEntries() + ", Size: "
+              + dfv.getSize());
+          i++;
+        }
+      }
+
+      log.debug("");
+      if (expectedFiles >= 0) {
```
[the remark this answers is not in the document]
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/util/FileMetadataUtil.java:69]: Yeah, -1 won't do the check and just prints. I was just trying to provide a quick way to print what's in metadata to see what is there to debug the test if not trying to actually count the files. The reason why this check is currently `>= 0` was I was thinking there might be a use case where you'd want to verify the expected files seen was in fact 0 (empty tablet) at some point in the future so if we made this change it wouldn't check that case of expected being 0 files. So I am thinking we should keep the current check as is so that case would work.

File: server/manager/src/main/java/org/apache/accumulo/manager/Manager.java [hunk 10]
```diff
@@ -649,16 +648,6 @@ TabletGoalState getGoalState(TabletLocationState tls, MergeInfo mergeInfo) {
               break;
             case STARTED:
             case SPLITTING:
```
[the remark this answers is not in the document]
Comment (cshannon) [reply server/manager/src/main/java/org/apache/accumulo/manager/Manager.java:650]: Yeah, this is actually already removed in my branch for #3769 which is about ready to be submitted as a PR as soon as this PR is merged. [Here](https://github.com/cshannon/accumulo/commit/54f03a443e55a35cb3bd951ab322fb027a90697a#diff-07984c23da1745b22dd4fde5c4b8877dfb190ea19e5e60ab79d8a93e752031c7L650) is the current line in the work in progress branch that would remove it

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java (hunk 5 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:1588]: That's a good point, do you think it makes sense to lazy cache the value after computation for repeat calls to `getLastKey()` on the same reader instance?

I haven't looked in detail but after a quick glance it appears that in the split code where `getLastKey()` is used it will just be called once but I'm wondering if there could be a use case in the future (or current that I missed) where the method could be called multiple times on the same reader so it would be worthwhile to avoid having to compute and copy the bytes each time.

File: core/src/main/java/org/apache/accumulo/core/gc/ReferenceFile.java (hunk 7 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/gc/ReferenceFile.java:70]: I think the javadoc here needs to be updated and I can do that. I made changes and forgot to update it. The path here could be exactly what's in metadata (such as if it's a `StoredTabletFile`) or may be normalized if the reference is `ScanServerRefTabletFile`.

I don't think it actually needs to be the exact path since we are not using it to remove entries in the metadata table and where the path is used later by `GarbageCollectionAlgorithm` it is actually processed and made relative anyways [here](https://github.com/cshannon/accumulo/blob/ecf7c7ac31379b5e5fc95a33df4237927edd8a9c/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java#L175)

File: server/manager/src/test/java/org/apache/accumulo/manager/TabletGroupWatcherTest.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    Pair<DataFileValue,DataFileValue> newValues = TabletGroupWatcher.computeNewDfv(original);
+
+    assertEquals(10, newValues.getFirst().getSize());
+    assertEquals(5, newValues.getFirst().getNumEntries());
+    assertEquals(original.getTime(), newValues.getFirst().getTime());
+    assertEquals(11, newValues.getSecond().getSize());
+    assertEquals(6, newValues.getSecond().getNumEntries());
+    assertEquals(original.getTime(), newValues.getSecond().getTime());
+  }
+
```
[the remark this answers is not in the document]
Comment (cshannon) [reply server/manager/src/test/java/org/apache/accumulo/manager/TabletGroupWatcherTest.java:54]: Added in https://github.com/apache/accumulo/pull/3761/commits/706c26702495373f0fa61a5ab6e1aba7b117bd0e

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java [hunk 12]
```diff
@@ -1556,5 +1561,230 @@ private void setInterruptFlagInternal(AtomicBoolean flag) {
...
+  public static RFileSKVIterator getReader(final CachableBuilder cb, final TabletFile dataFile)
+      throws IOException {
+    final RFile.Reader reader = new RFile.Reader(Objects.requireNonNull(cb));
+    return dataFile.hasRange() ? new FencedReader(reader, dataFile.getRange()) : reader;
+  }
+
+  public static RFileSKVIterator getReader(final CachableBuilder cb, Range range)
+      throws IOException {
+    final RFile.Reader reader = new RFile.Reader(Objects.requireNonNull(cb));
+    return !Objects.requireNonNull(range).isInfiniteStartKey() || !range.isInfiniteStopKey()
```
[the remark this answers is not in the document]
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java]: Fixed in https://github.com/apache/accumulo/pull/3761/commits/706c26702495373f0fa61a5ab6e1aba7b117bd0e

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java [hunk 13]
```diff
@@ -76,6 +78,38 @@ public interface InputArguments {
...
+    /**
+     * @since 3.1.0
+     */
+    class FencedPath {
+      private final Path path;
+      private final Range fence;
+
+      public FencedPath(Path path, Range fence) {
+        this.path = Objects.requireNonNull(path);
+        this.fence = Objects.requireNonNull(fence);
```
[the remark this answers is not in the document]
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java]: I added this and added tests in https://github.com/apache/accumulo/pull/3761/commits/706c26702495373f0fa61a5ab6e1aba7b117bd0e

**Comment by cshannon** [comment]
@keith-turner - I added the 4 tests you suggested to the list here https://github.com/apache/accumulo/issues/3766 . I figured the simplest thing to do was just have 1 issue that tracks all the tests to add and create PRs for each one so we don't end up with a ton of separate issues.

Your external compaction test suggestion made me think of something else. I realized that in some of the tests for Merge/Delete I am counting and verifying the metadata files are the expected count to make sure things look ok in the table. I realized a compaction running could throw this off so I am wondering if I need to update and make sure the ITs won't ever run a compaction or if it's not necessary as there's not enough data being written. Probably wouldn't hurt to disable just so things are deterministic.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java (hunk 5 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:1588]: @keith-turner - another option I thought of was just caching using Guava memoize which we've done elsewhere. I created a separate branch/commit as a quick example of how this would look and wanted to see what you thought: https://github.com/cshannon/accumulo/commit/d871e466ac90a02dfd78fdbda90ca4a1b0d8cfa7

File: core/src/main/java/org/apache/accumulo/core/gc/ReferenceFile.java (hunk 7 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/gc/ReferenceFile.java:70]: > I don't think it actually needs to be the exact path since we are not using it to remove entries in the metadata table and where the path is used later by GarbageCollectionAlgorithm it is actually processed and made relative anyways [here](https://github.com/cshannon/accumulo/blob/ecf7c7ac31379b5e5fc95a33df4237927edd8a9c/server/gc/src/main/java/org/apache/accumulo/gc/GarbageCollectionAlgorithm.java#L175)

I agree with that and saw the same thing when I looked.  Based on that it does not seem we need the exact bytes that are stored in the metadata table.

File: core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java (hunk 5 above)
Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:1588]: > another option I thought of was just caching using Guava memoize which we've done elsewhere. 

I like that solution, multiple benefits w/o any downside.

File: test/src/main/java/org/apache/accumulo/test/util/FileMetadataUtil.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply test/src/main/java/org/apache/accumulo/test/util/FileMetadataUtil.java:69]: > I was thinking there might be a use case where you'd want to verify the expected files seen was in fact 0

Hah, that is what I wanted to achieve too and the code already did that.  I think I made the right call to stop reviewing this yesterday even though I had not gone through all the files.

**Comment by keith-turner** [comment]
> I realized a compaction running could throw this off so I am wondering if I need to update and make sure the ITs won't ever run a compaction or if it's not necessary as there's not enough data being written. Probably wouldn't hurt to disable just so things are deterministic.

Yeah it is nice to avoid non deterministic test failures.  Could up the compaction ratio on the table to higher than the number of files created to avoid this.   If creating less files than the default for the compaction ratio it should not be a problem.

File: core/src/test/java/org/apache/accumulo/core/file/rfile/FencedRFileTest.java [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+import org.apache.accumulo.core.file.FileSKVIterator;
+import org.apache.accumulo.core.file.rfile.RFile.FencedReader;
+import org.apache.accumulo.core.iterators.SortedKeyValueIterator;
+import org.apache.accumulo.core.iteratorsImpl.system.MultiIterator;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.Text;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class FencedRFileTest extends AbstractRFileTest {
```
Comment (keith-turner) [inline core/src/test/java/org/apache/accumulo/core/file/rfile/FencedRFileTest.java:43]: Ran these unit test w/ code coverage and noticed some of the new code was not covered.  Added an item to the list on #3766 about this.

File: core/src/main/java/org/apache/accumulo/core/metadata/ReferencedTabletFile.java [hunk 15]
```diff
@@ -186,48 +195,35 @@ public String getNormalizedPathStr() {
...
   }
 
   @Override
   public int compareTo(ReferencedTabletFile o) {
     if (equals(o)) {
       return 0;
     } else {
-      return parts.getNormalizedPath().compareTo(o.parts.getNormalizedPath());
+      return comparator.compare(this, o);
     }
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/ReferencedTabletFile.java:211]: This was an existing pattern in the code, may be better to skip the equals call.  In the not equals case it will read the data twice.

**Review by keith-turner — APPROVED** [review]

File: core/src/test/java/org/apache/accumulo/core/file/rfile/FencedRFileTest.java (hunk 14 above)
Comment (cshannon) [reply core/src/test/java/org/apache/accumulo/core/file/rfile/FencedRFileTest.java:43]: This may be similar to https://github.com/apache/accumulo/issues/3448 that was opened, I went ahead and just linked that and added it to the list i #3766

File: core/src/main/java/org/apache/accumulo/core/metadata/ReferencedTabletFile.java (hunk 15 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/ReferencedTabletFile.java:211]: Good point, i can drop the equals for sure, just an oversight, good catch. It should just be:

```java
  @Override
  public int compareTo(ReferencedTabletFile o) {
    return comparator.compare(this, o);
  }
```

**Comment by cshannon** [comment]
I just kicked off one more full IT and assuming that passes I am planning to merge this today. There are several follow on issues to complete and they can't be submitted until this is merged. 3.1 is not going to be released for a while so there is plenty of time to put in follow on issues and PRs to address any issues with no-chop merge that comes up during testing or later before release.

**Comment by cshannon** [comment]
> > I realized a compaction running could throw this off so I am wondering if I need to update and make sure the ITs won't ever run a compaction or if it's not necessary as there's not enough data being written. Probably wouldn't hurt to disable just so things are deterministic.
> 
> Yeah it is nice to avoid non deterministic test failures. Could up the compaction ratio on the table to higher than the number of files created to avoid this. If creating less files than the default for the compaction ratio it should not be a problem.

Looks like I did do this in one [test](https://github.com/cshannon/accumulo/blob/3673ff8c297bdccac9bc0f55fb726244e8157d7d/test/src/main/java/org/apache/accumulo/test/functional/MergeIT.java#L201) but need to go back and make sure other tests have it too, will add it to  #3766

**Comment by cshannon** [comment]
> I just kicked off one more full IT and assuming that passes I am planning to merge this today. There are several follow on issues to complete (including a major one of merging into elasticity) and they can't be completed until this is merged. 3.1 is not going to be released for a while so there is plenty of time to put in follow on issues and PRs to address any issues with no-chop merge that comes up during testing or later before release.

The only thing that failed was GarbageCollectorIT due to updates from #3738 which have now been fixed
